### PR TITLE
Clean up CoAP CLI.

### LINKED ISSUE #1863: cli coap example does not seem to work
make -f examples/Makefile-posix

In one terminal window:
```
$ ./ot-cli-ftd 1
> 
> networkname foobar
Done
> panid 0x3333
Done
> channel 21
Done
> ifconfig up
Done
> thread start
Done
> state
child
Done
> ipaddr
fdde:ad00:beef:0:0:ff:fe00:c01
fdde:ad00:beef:0:2f58:72b1:818a:8ef3
fe80:0:0:0:10af:5b24:f29c:2c79
Done
>  
> coap server name Foo
Changing resource name to 'Foo': Done
> coap server start
Server started with resource 'Foo': Done
> Received CoAP request from [fe80:0:0:0:10af:5b24:f29c:2c79]: GET
CoAP response sent successfully!
Received CoAP request from [fe80:0:0:0:10af:5b24:f29c:2c79]: PUT with payload: 48 69 4d 6f 6d 
CoAP response sent successfully!
Received CoAP request from [fe80:0:0:0:10af:5b24:f29c:2c79]: GET
CoAP response sent successfully!
```
In a second terminal window
```
$ ./ot-cli-ftd 2

> 
> networkname foobar
Done
> panid 0x3333
Done
> channel 21
Done
> ifconfig up
Done
> thread start
Done
> state
leader
Done
> 
> ipaddr
fdde:ad00:beef:0:0:ff:fe00:fc00
fdde:ad00:beef:0:0:ff:fe00:c00
fdde:ad00:beef:0:44ff:f203:bed5:dd89
fe80:0:0:0:bc4c:f66:3bb4:9c00
Done
> 
> coap client get fe80:0:0:0:10af:5b24:f29c:2c79 Foo                        
Sending CoAP message: Done
> Error receiving CoAP response message: 28
```
Note: Over the air, using real hardware and a sniffer, I see the CoAP response.
I also tried other variants.
```
> coap client put fe80:0:0:0:10af:5b24:f29c:2c79 Foo con HiMom
Sending CoAP message: Done
> 
> coap client get fe80:0:0:0:10af:5b24:f29c:2c79 Foo con
Sending CoAP message: Done
> 
> Error receiving CoAP response message: 28
Error receiving CoAP response message: 28
```
This contradicts the example in the documentation.
```
coap client get fdde:ad00:beef:0:dbaa:f1d0:8afb:30dc test
Sending CoAP message: Done
Received CoAP request from [fdde:ad00:beef:0:dbaa:f1d0:8afb:30dc]: GET
CoAP response sent successfully!
```
Am I missing something?

### PR DESCRIPTION
**Pull request by jwhui**
- Change static methods into class methods.
- Remove client/server distinction.
- Fix callback method types and remove unnecessary type casts.
- Update README.md to reflect changes.

Resolves #1863.

### CODE REVIEW

**Comment by DuaneEllis-TI** [comment]
>> Fix callback method types and remove unnecessary type casts.

Huh??   I don't see something that fixes #1863 

I see stuff that cleans the API up... but not the problem that I am reporting.

Can you point me to the part that fixes that issue?

**Comment by jwhui** [comment]
You need to start the coap service on both client and server sides.  The prior CLI was confusing because only `coap server start` existed, but not `coap client start`.  At the same time, the current coap implementation no longer splits between server and client.  As a result, the cli has been collapsed down to `coap start`.

**Comment by DuaneEllis-TI** [comment]
yea, that's confusing...  thanks it works now.

File: src/cli/cli_coap.cpp [hunk 1]
```diff
@@ -83,109 +72,74 @@ void Coap::PrintPayload(otMessage *aMessage)
...
+    mServer->OutputFormat("\r\n");
 }
 
-void Coap::OutputBytes(const uint8_t *aBytes, uint8_t aLength)
+void Coap::OutputBytes(const uint8_t *aBytes, uint8_t aLength) const
 {
     for (int i = 0; i < aLength; i++)
     {
-        sServer->OutputFormat("%02x ", aBytes[i]);
+        mServer->OutputFormat("%02x ", aBytes[i]);
```
Comment (bukepo) [inline src/cli/cli_coap.cpp]: Remove the trailing space in `%02x ` to keep consistent with `Interpreter::OutputBytes`, or just reuse it?

Comment (DuaneEllis-TI) [inline src/cli/cli_coap.cpp]: Why not use Interpreter::OutputBytes instead

Comment (jwhui) [inline src/cli/cli_coap.cpp]: We can.  I'll need to update the code to maintain a reference to the `Interpreter` object.

**Pull request merged**